### ISSUE TITLE
npm v7 support

### ISSUE THREAD

**Issue by JamieMagee** [renovatebot/renovate#7474]
**What would you like Renovate to be able to do?**

Support the newly released npm v7. The blog post [Presenting v7.0.0 of the npm CLI](https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/) goes over some of the new features and breaking changes.

I'd like to use this issue as a mega-issue to discuss and co-ordinate npm v7 related changes to Renovate.

**Did you already have any implementation ideas?**

Some of the key features that will require Renovate changes:

- [ ] `package-lock.json` v2

    I can't yet find a schema for this, but [here's a good example](https://github.com/npm/cli/blob/latest/package-lock.json)

- [ ] Workspaces

    See [RFC 0026](https://github.com/npm/rfcs/blob/latest/implemented/0026-workspaces.md)

- [ ] `npx`

    > npx has been completely rewritten to use the npm exec command. There are various changes in functionality, most noticeable being a prompt if the module you are trying to run is not yet installed.

    There are a few places we use `npx`, and we should make sure that we are calling `npx` non-interatively.

- [ ] Peer dependencies

    Not sure if this one affects us

    > Automatically installing peer dependencies (while this feature is something we think is desirable new behavior, it does potentially break certain workflows).

This list may not be complete.

The new library, [`@npmcli/arborist`](https://www.npmjs.com/package/@npmcli/arborist) is also quite interesting, and should make interacting with the `node_modules` tree easier in future.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

Switch to yarn :joy:

**Is this a feature you'd be interested in implementing yourself?**

Yes, with assistance

**Comment by mccxiv** [comment]
I'm here to express my confusion because we're already receiving PRs from renovate with `"lockfileVersion": 2`.  
Is this already implemented?

**Comment by rarkins** [comment]
@mccxiv it's not intended. Can you reproduce the problem in a public repo? Does your engines allow node 15?

**Comment by mccxiv** [comment]
I don't have a public repro for now

We use a monorepo on Gitlab. The top level `package.json` has: 
```
"engines": {
    "node": ">= 10.0.0",
    "npm": ">= 5.6.0"
  },
```
The individual packages do not have an `engines:` field

The odd thing is some PRs don't use lockfile v2, and some do (see line diffs):

![](https://puu.sh/GHd8W/3967e4ce9a.png)

![](https://puu.sh/GHd94/fb3ea5a136.png)

![](https://puu.sh/GHd9j/d8669bf823.png)

**Comment by Quramy** [comment]
> we're already receiving PRs from renovate with "lockfileVersion": 2

@rarkins  I have same issue.

Here is my reproducing repo: https://github.com/Quramy/npm-ts-workspaces-example/pull/3#issuecomment-716925358 (My repo uses NPM v7 workspaces feature too).

[FYI]

I found https://github.com/renovatebot/renovate/blob/master/lib/manager/npm/post-update/npm.ts#L30-L33 

And I configured `constraints` params in order to tell that my NPM version is v7, but the bot still outputs update issue 🤔 

```json
{
  "extends": ["config:base"],
  "constraints": {
    "npm": "^7.0.0"
  }
}
```

**Comment by rarkins** [comment]
@Quramy what is your issue? You constraints say npm 7 so you're going to get lockfileversion 2.

**Comment by rarkins** [comment]
I can reproduce it. If you have an engines that allows `node` 15, and no `npm` constraint limiting <7, then Renovate will run `npm install` using Node 15 which defaults to npm 7. Example log:

```
DEBUG: Found compatible npm version (repository=renovate-tests/npm42, branch=renovate/chalk-2.x)
       "constraint": ">=12",
       "version": "15.0.1"
DEBUG: Resolved tag constraint (repository=renovate-tests/npm42, branch=renovate/chalk-2.x)
       "image": "docker.io/renovate/node",
       "tagConstraint": ">=12",
       "tagVersioning": "npm",
       "tag": "15.0.1"
```

**Comment by rarkins** [comment]
I've experimented with using an npm constraint (e.g. `npm i npm@^6.0.0`) but it doesn't seem to downgrade the `npm` version in `node 15`.

**Comment by rarkins** [comment]
If you add the following to your `renovate.json` then it succeeds in getting Renovate to use node 14:

```
  "constraints": {
    "node": "< 15.0.0"
  }
```

Unfortunately it's only succeeding due to what seems like a *bug*, but it can work for now and be forwards compatible for anyone who wants to support node 15 in their engines without running it during Renovate PRs.

**Comment by Quramy** [comment]
Sorry, the bot update failure in my repo was not Renovate issue. Probably because some of the dependencies couldn't be installed in the Node v15 environment and I misunderstood it as package-lock version issue :bow:

**Comment by rarkins** [comment]
Actually, the required config will be as following once #7561 is merged and live:

```
{
  "force": {
    "constraints": {
      "node": "< 15.0.0"
    }
  }
}
```

The `force` wrapper is necessary because otherwise the detected engines node constraint takes precedence.

**Comment by ylemkimon** [comment]
@rarkins 

> I've experimented with using an npm constraint (e.g. `npm i npm@^6.0.0`) but it doesn't seem to downgrade the `npm` version in `node 15`.

Do you mean by setting
```
{
  "constraints": {
    "npm": "^6.0.0"
  }
}
```
?

In that case, it would haven't worked due to the cached location of npm after installing npm:
https://github.com/renovatebot/renovate/blob/367e591c66cd071257966b05a1ca9a63849ab8df/lib/manager/npm/post-update/npm.ts#L30-L33

I opened a PR to fix the issue including automatically setting the `constraints` in #7700.

**Comment by ylemkimon** [comment]
> * [ ]  Workspaces
>   See [RFC 0026](https://github.com/npm/rfcs/blob/latest/implemented/0026-workspaces.md)

Workspaces seem to work out of the box: https://github.com/ylemkimon/npm-workspace.

> * [ ]  `npx`
>   > npx has been completely rewritten to use the npm exec command. There are various changes in functionality, most noticeable being a prompt if the module you are trying to run is not yet installed.
>   
>   
>   There are a few places we use `npx`, and we should make sure that we are calling `npx` non-interatively.

The only place `npx` is used is: https://github.com/renovatebot/renovate/blob/75737805c804f69f427e0f8147c463836b1af448/lib/manager/npm/post-update/yarn.ts#L145-L156

**Comment by ylemkimon** [comment]
`npx` will also work out of the box, i.e., non-interactively, because we set `CI=true`:
https://github.com/renovatebot/renovate/blob/75737805c804f69f427e0f8147c463836b1af448/lib/manager/npm/post-update/yarn.ts#L64-L67
and npm 7 detects it: https://github.com/npm/cli/blob/a28aff769a77f127f371c31afcb9e9814722e5cd/lib/exec.js#L190-L191

I think we can say Renovate has full support for npm 7.

**Comment by stephenwade** [comment]
#7700 adds support for npm@7 in Docker, but npm@6 is still used when run locally (`npm i -g renovate`). Is there any way to get Renovate to use npm@7 (or get it to use my globally installed npm instead of the bundled npm) when running locally?

**Comment by viceice** [comment]
Looks like this is a technical limitation, as yarn / npm always prefer local bin before global.

Maybe we need to migrate away from referencing the full npm package.

**Comment by stephenwade** [comment]
Renovate doesn't reference npm directly, but semantic-release does so it still ends up in the local bin folder.

**Comment by stephenwade** [comment]
Are there any config options to specify the path to npm?

**Comment by viceice** [comment]
Semantic release is a d v dependency and shouldn't be installed on production. 🤔

**Comment by viceice** [comment]
> Are there any config options to specify the path to npm?

Nope

**Comment by stephenwade** [comment]
Okay, I found my issue. I was testing with a local dev build which is why I had npm@6 in that folder. When I run the globally installed renovate, it uses my globally installed npm version.

**Comment by ylemkimon** [comment]
@stephenwade Try using [`--binary-source=docker`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource).

**Comment by HonkingGoose** [comment]
@rarkins Should we maybe bump the priority label on this to `priority-2-important` now that npm 7 is `latest`?

**Comment by viceice** [comment]
I think we properly support npm v7 now. 🤔

I'll close this, feel free to reopen or open a new issue / discussion if there are any problems.

**Comment by rarkins** [comment]
I think we're missing workspaces, but let's handle that in a separate issue and get a reproduction for it

**Comment by abernix** [comment]
🤔 It seems that, despite having `engines.npm` set to a 6.x range, that the attempt to update is still happening with npm 7:

From https://github.com/apollographql/apollo-server/pull/4998#issuecomment-792746046:

```
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: undefined,
npm WARN EBADENGINE   required: { node: '>=6 <15', npm: '^6.14.11' },
npm WARN EBADENGINE   current: { node: 'v14.16.0', npm: '7.6.1' }
npm WARN EBADENGINE }
```

Is there another way that this is meant to be configured/conveyed?

**Comment by rarkins** [comment]
@abernix please create a new bug report issue and locate the relevant job logs from `app.renovatebot.com`. We'd want to see any mentions of `constraints` in the logs as well as see the debugging prior to the `docker run` command being executed.

**Comment by abernix** [comment]
@rarkins I tried checking the logs on the dashboard, but it doesn't appear to be loading right now for me: https://app.renovatebot.com/dashboard (I tried logging out, but now I can't log in again)

**Comment by HonkingGoose** [comment]
@rarkins The Renovate Dashboard won't let me login either. Do you want a new issue from me to track this? Or is this something internal?

**Comment by rarkins** [comment]
Can you try again now? I restarted the API process and I think that's cleared it

**Comment by abernix** [comment]
@rarkins Dashboard is good now! Thanks.

**Comment by gremo** [comment]
Hi, for me "force" seems not to work. 

Renovate is still creating PR with NPM version 7 (see https://github.com/gremo/nest-winston/commit/4f8ddb801da9c886dd8f642748b8ef1e5f9f92bf). Any help is much appreciated!

```json
{
  "extends": [
    "config:base",
    ":disablePeerDependencies"
  ],
  "packageRules": [
    {
      "depTypeList": [
        "devDependencies"
      ],
      "rangeStrategy": "bump"
    }
  ],
  "force": {
    "constraints": {
      "node": "< 15.0.0"
    }
  }
}
```

**Comment by JamieMagee** [comment]
@gremo Can you please open a new issue, or comment on an existing open issue. @abernix's issue #9036 appears to be relevant.

**Comment by gremo** [comment]
@JamieMagee thanks created a new issue https://github.com/renovatebot/renovate/issues/9103. Hope I find where the problem is.

**Comment by rarkins** [comment]
It's the same scenario as for @abernix - forced constraints with node only now don't stop npm@latest from being installed.

**Comment by gremo** [comment]
Thanks @rarkins, so... what's the proposed / working solution?

**Comment by viceice** [comment]
@gremo Do not use the force constraints and let the renovatew magic work.

Als long there is a v1 lockfile, renovate will force node v14 and npm v6.
